include/metaprogram.h

```diff
@@ -81,6 +82,18 @@
 #define R_FOR_EACH_WITH_(macro, args, a, ...) INVOKE_WITH(macro, args, a) __VA_OPT__(R_FOR_EACH_WITH_P PARENS (macro, args, __VA_ARGS__))
 #define R_FOR_EACH_WITH_P() R_FOR_EACH_WITH_
 
+/* Expands to 'macro(a, b)' for each 'a' in 'as' and 'b' in 'bs'.
+ * Uses the shorter of 'as' and 'bs'. */
+#define R_ZIP_WITH(macro, as, bs) CAT(R_ZIP_WITH_, CAT(R_ZIP_WITH_NONEMPTY(as), R_ZIP_WITH_NONEMPTY(bs)))(macro, FIRST as, FIRST bs, (EXCEPT_1 as), (EXCEPT_1 bs))
+#define R_ZIP_WITH_00(macro, a, b, as, bs)
+#define R_ZIP_WITH_01(macro, a, b, as, bs)
+#define R_ZIP_WITH_10(macro, a, b, as, bs)
+#define R_ZIP_WITH_11(macro, a, b, as, bs) macro(a, b) R_ZIP_WITH_P PARENS (macro, as, bs)
+#define R_ZIP_WITH_P() R_ZIP_WITH
+
+#define R_ZIP_WITH_NONEMPTY(as) R_ZIP_WITH_NONEMPTY_ as
+#define R_ZIP_WITH_NONEMPTY_(...) FIRST(__VA_OPT__(1,) 0)
```

Comment:
```suggestion
#define R_ZIP(macro, as, bs) CAT(R_ZIP_, CAT(R_ZIP_NONEMPTY(as), R_ZIP_NONEMPTY(bs)))(macro, FIRST as, FIRST bs, (EXCEPT_1 as), (EXCEPT_1 bs))
#define R_ZIP_00(macro, a, b, as, bs)
#define R_ZIP_01(macro, a, b, as, bs)
#define R_ZIP_10(macro, a, b, as, bs)
#define R_ZIP_11(macro, a, b, as, bs) macro(a, b) R_ZIP_P PARENS (macro, as, bs)
#define R_ZIP_P() R_ZIP

#define R_ZIP_NONEMPTY(as) R_ZIP_NONEMPTY_ as
#define R_ZIP_NONEMPTY_(...) FIRST(__VA_OPT__(1,) 0)
```
I realized that this should be called `R_ZIP` not `R_ZIP_WITH`:
- `R_FOR_EACH` does `m(a0) m(a1) ... m(aN)`
- `R_FOR_EACH_WITH` does `m(...args, a0) m(...args, a1) ... m(...args, aN)`
- So `R_ZIP` should do `m(a0, b0) m(a1, b1) ... m(aN, bN)`
- And then a hypothetical `R_ZIP_WITH` would also allow you to provide some arguments (we can add this in the future if we need it).

~that mrgriffin guy who PRed it to your branch is an idiot~